EmailValidator/Validation/MailboxCheckValidation.php

```diff
@@ -0,0 +1,240 @@
+
+class MailboxCheckValidation implements EmailValidation
+{
+    const END_OF_LINE = "\r\n";
+
+    /**
+     * @var InvalidEmail
+     */
+    private $error;
+
+    /**
+     * @var Warning[]
+     */
+    private $warnings = [];
+
+    /**
+     * @var int
+     */
+    private $lastResponseCode;
+
+    /**
+     * @var int
+     */
+    private $port = 25;
+
+    /**
+     * @var int
+     */
+    private $timeout = 10;
+
+    /**
+     * @var string
+     */
+    private $fromEmail = 'test-mailbox@validation.email';
+
+    /**
+     * MailboxCheckValidation constructor.
+     */
+    public function __construct()
+    {
+        if (!extension_loaded('intl')) {
+            throw new \LogicException(sprintf('The %s class requires the Intl extension.', __CLASS__));
+        }
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getError()
+    {
+        return $this->error;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getWarnings()
+    {
+        return $this->warnings;
+    }
+
+    /**
+     * @return int
+     */
+    public function getLastResponseCode()
+    {
+        return $this->lastResponseCode;
+    }
+
+    /**
+     * @return int
+     */
+    public function getPort()
+    {
+        return $this->port;
+    }
+
+    /**
+     * @param int $port
+     */
+    public function setPort($port)
+    {
+        $this->port = $port;
+    }
+
+    /**
+     * @return int
+     */
+    public function getTimeout()
+    {
+        return $this->timeout;
+    }
+
+    /**
+     * @param int $timeout
+     */
+    public function setTimeout($timeout)
+    {
+        $this->timeout = $timeout;
+    }
+
+    /**
+     * @return string
+     */
+    public function getFromEmail()
+    {
+        return $this->fromEmail;
+    }
+
+    /**
+     * @param string $fromEmail
+     */
+    public function setFromEmail($fromEmail)
+    {
+        $this->fromEmail = $fromEmail;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function isValid($email, EmailLexer $emailLexer)
+    {
+        $mxHosts = $this->getMXHosts($email);
+
+        $isValid = false;
+        foreach ($mxHosts as $mxHost) {
+            if ( ($isValid = $this->checkMailbox($mxHost, $this->port, $this->timeout, $this->fromEmail, $email)) ) {
+                break;
+            }
+        }
+
+        if ( ! $isValid ) {
+            $this->error = new IllegalMailbox($this->lastResponseCode);
+        }
+
+        return $this->error === null;
+    }
+
+    /**
+     * @param string $email
+     *
+     * @return array
+     */
+    protected function getMXHosts($email)
+    {
+        $variant = defined('INTL_IDNA_VARIANT_UTS46') ? INTL_IDNA_VARIANT_UTS46 : INTL_IDNA_VARIANT_2003;
+
+        $hostname = $email;
+        if ( false !== ($lastAtPos = strrpos($email, '@')) ) {
+            $hostname = substr($email, $lastAtPos + 1);
+        }
+        $hostname = rtrim(idn_to_ascii($hostname, IDNA_DEFAULT, $variant), '.') . '.';
```

Comment:
Please avoid mutating variables this way, it is hard to follow.
Maybe
```php
$hostname = $this->extractHostname($email);

private function extractHostname($email)
{
  $lastAtPos = (bool) strrpos($email, '@')
  $variant = defined('INTL_IDNA_VARIANT_UTS46') ? INTL_IDNA_VARIANT_UTS46 : INTL_IDNA_VARIANT_2003;
  if ($lastAtPos) { 
    $removedAT = substr($email, $lastAtPos + 1);
    return rtrim(idn_to_ascii($removedAT, IDNA_DEFAULT, $variant), '.') . '.';
  }
  return rtrim(idn_to_ascii($email, IDNA_DEFAULT, $variant), '.') . '.';
}
```

Reply:
fix